File: text_forge_backend/models/story_flow.py
"""剧情流（交互式章节剧情推演）数据模型"""
from datetime import datetime

from sqlalchemy import (
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from models.base import Base


class StoryFlow(Base):
    """剧情流会话表：一次交互式章节剧情推演对应一条会话。"""

    __tablename__ = "story_flows"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True, comment="剧情流ID")
    book_id: Mapped[int] = mapped_column(ForeignKey("books.id", ondelete="CASCADE"), nullable=False, index=True, comment="所属书籍ID")
    chapter_id: Mapped[int] = mapped_column(ForeignKey("chapters.id", ondelete="CASCADE"), nullable=False, index=True, comment="所属章节ID")
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True, comment="创建用户ID")
    status: Mapped[str] = mapped_column(String(20), default="active", comment="会话状态: active/completed")
    anchor_event_ids: Mapped[list] = mapped_column(JSONB, default=[], comment="锚点事件ID快照（生成时锁定；无事件时为空列表）")
    current_event_index: Mapped[int] = mapped_column(Integer, default=0, comment="当前推进到的事件下标，-1 表示实时生成模式")
    view_character_id: Mapped[int] = mapped_column(Integer, nullable=True, comment="视角角色ID（不加DB外键，角色可能被删除，应用层校验）")
    round_count: Mapped[int] = mapped_column(Integer, default=0, comment="已生成的场景数")
    summary: Mapped[str] = mapped_column(Text, nullable=True, comment="推演摘要（提交agent时生成，留档）")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), comment="创建时间")
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now(), comment="更新时间")

    nodes: Mapped[list["StoryFlowNode"]] = relationship(back_populates="flow", cascade="all,delete-orphan")


class StoryFlowNode(Base):
    """剧情流场景节点表：一个场景（叙事+选项）对应一行，决策链由此派生。"""

    __tablename__ = "story_flow_nodes"
    __table_args__ = (
        UniqueConstraint("flow_id", "seq", name="uq_story_flow_nodes_flow_seq"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True, comment="节点ID")
    flow_id: Mapped[int] = mapped_column(ForeignKey("story_flows.id", ondelete="CASCADE"), nullable=False, index=True, comment="所属剧情流ID")
    seq: Mapped[int] = mapped_column(Integer, nullable=False, comment="节点序号（1 起，flow_id+seq 联合唯一防重复插入）")
    anchored_event_id: Mapped[int] = mapped_column(Integer, nullable=True, comment="锚定的事件ID，实时生成模式为 null")
    title: Mapped[str] = mapped_column(String(200), nullable=False, comment="场景标题")
    narration: Mapped[str] = mapped_column(Text, nullable=False, comment="叙事文本（LLM 生成原文，含视角角色原名）")
    options: Mapped[list] = mapped_column(JSONB, default=[], comment="预设选项数组 [{\"text\": ...}]")
    chosen_option: Mapped[str] = mapped_column(Text, nullable=True, comment="用户选择的选项文本（自定义输入原样存）")
    location_name: Mapped[str] = mapped_column(String(200), nullable=True, comment="场景地点名（展示用）")
    character_names: Mapped[list] = mapped_column(JSONB, default=[], comment="出场角色名列表（展示用）")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), comment="创建时间")

    flow: Mapped["StoryFlow"] = relationship(back_populates="nodes")

File: text_forge_front_remastered_edition/src/shared/api/storyFlow.ts
// 剧情流 REST/SSE 封装，统一对接后端 /api/story-flows 接口。
// SSE 事件命名与 agent 流隔离：scene_stream / scene_done / done / error。
import { apiClient } from './client';
import { authedFetch } from './authFetch';
import { readSSE } from './sse';

export interface StoryFlowOption {
  text: string;
}

export interface StoryFlowNode {
  id: number;
  seq: number;
  anchoredEventId: number | null;
  title: string;
  narration: string;
  options: StoryFlowOption[];
  chosenOption: string | null;
  locationName: string | null;
  characterNames: string[];
  createdAt?: string;
}

export interface StoryFlow {
  id: number;
  bookId: number;
  chapterId: number;
  status: 'active' | 'completed';
  anchorEventIds: number[];
  currentEventIndex: number;
  viewCharacterId: number | null;
  roundCount: number;
  summary?: string | null;
  createdAt?: string;
  updatedAt?: string;
}

export interface StoryFlowFetchResult {
  flow: StoryFlow;
  nodes: StoryFlowNode[];
}

export interface StoryFlowStreamCallbacks {
  onStream?: (token: string) => void;
  onSceneDone?: (data: {
    node: StoryFlowNode | null;
    completed: boolean;
    flowId: number;
    anchorEventIds?: number[];
    currentEventIndex?: number;
  }) => void;
  onDone?: () => void;
  onError?: (message: string) => void;
}

export interface StoryFlowModelConfig {
  main_config?: { adapter?: string; base_url?: string; api_key?: string; model_id?: string };
  search_config?: { provider?: string; api_key?: string };
}

export class StoryFlowConfigError extends Error {}

/** 校验模型配置已设置（未配置时抛出明确错误，UI 据此 toast）。 */
export function assertStoryFlowModelConfig(modelConfigData: StoryFlowModelConfig | null): StoryFlowModelConfig {
  if (!modelConfigData || !modelConfigData.main_config) {
    throw new StoryFlowConfigError('请先在设置页配置模型');
  }
  return modelConfigData;
}

/** 读取 SSE 流并按事件回调分发（复用 agent.ts streamAgent 的 reader/decoder 模式）。 */
const API_BASE = process.env.NEXT_PUBLIC_API_URL || '/api';

async function streamStoryFlow(
  url: string,
  body: Record<string, unknown>,
  callbacks: StoryFlowStreamCallbacks,
  abortSignal?: AbortSignal,
): Promise<void> {
  const res = await authedFetch(`${API_BASE}${url}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: abortSignal,
  });

  if (!res.ok) {
    let detail = '剧情流请求失败';
    try {
      const d = await res.json();
      if (d?.detail) detail = typeof d.detail === 'string' ? d.detail : detail;
    } catch { /* ignore */ }
    callbacks.onError?.(detail);
    return;
  }

  await readSSE(res, (event) => {
    switch (event.type) {
      case 'scene_stream':
        if (typeof event.token === 'string') callbacks.onStream?.(event.token);
        break;
      case 'scene_done':
        callbacks.onSceneDone?.({
          node: (event.node as StoryFlowNode | null) ?? null,
          completed: Boolean(event.completed),
          flowId: Number(event.flow_id ?? 0),
          anchorEventIds: Array.isArray(event.anchor_event_ids) ? (event.anchor_event_ids as number[]) : undefined,
          currentEventIndex: typeof event.current_event_index === 'number' ? event.current_event_index : undefined,
        });
        break;
      case 'done':
        callbacks.onDone?.();
        break;
      case 'error':
        callbacks.onError?.(typeof event.message === 'string' ? event.message : '生成失败，请重试');
        break;
    }
  });
}

/** 创建剧情流会话并流式生成首场景（幂等：已有 active 流则复用）。 */
export async function createStoryFlow(
  bookId: number,
  chapterId: number,
  viewCharacterId: number | null,
  modelConfigData: StoryFlowModelConfig | null,
  callbacks: StoryFlowStreamCallbacks,
  abortSignal?: AbortSignal,
): Promise<void> {
  assertStoryFlowModelConfig(modelConfigData);
  await streamStoryFlow(
    '/story-flows/',
    {
      bookId,
      chapterId,
      viewCharacterId: viewCharacterId ?? null,
      modelConfig: modelConfigData,
    },
    callbacks,
    abortSignal,
  );
}

/** 推进剧情流（幂等：已生成的后续节点直接回放，不重复生成）。 */
export async function advanceStoryFlow(
  flowId: number,
  chosenOption: string,
  modelConfigData: StoryFlowModelConfig | null,
  callbacks: StoryFlowStreamCallbacks,
  abortSignal?: AbortSignal,
): Promise<void> {
  assertStoryFlowModelConfig(modelConfigData);
  await streamStoryFlow(
    `/story-flows/${flowId}/advance`,
    { chosenOption, modelConfig: modelConfigData },
    callbacks,
    abortSignal,
  );
}

/** 结束剧情流并生成推演摘要（幂等）。 */
export async function completeStoryFlow(
  flowId: number,
  modelConfigData: StoryFlowModelConfig | null,
): Promise<{ summary: string; status: string; flowId: number }> {
  assertStoryFlowModelConfig(modelConfigData);
  const { data } = await apiClient.post<{ summary: string; status: string; flowId: number }>(
    `/story-flows/${flowId}/complete`,
    { modelConfig: modelConfigData },
  );
  return data;
}

/** 获取剧情流会话与全部节点（按 seq 有序，恢复用）。 */
export async function fetchStoryFlow(flowId: number): Promise<StoryFlowFetchResult> {
  const { data } = await apiClient.get<StoryFlowFetchResult>(`/story-flows/${flowId}`);
  return data;
}

/** 查询章节下未完成的剧情流（自动恢复用；返回 updated_at 最新一条或 null）。 */
export async function findActiveStoryFlow(bookId: number, chapterId: number): Promise<StoryFlow | null> {
  const { data } = await apiClient.get<{ items: StoryFlow[] }>('/story-flows/', {
    params: { bookId, chapterId, status: 'active' },
  });
  return (data.items ?? [])[0] ?? null;
}

/** 更新剧情流视角角色。 */
export async function updateStoryFlowViewCharacter(
  flowId: number,
  viewCharacterId: number | null,
): Promise<StoryFlow> {
  const { data } = await apiClient.patch<StoryFlow>(`/story-flows/${flowId}`, {
    viewCharacterId,
  });
  return data;
}

/** 查询剧情流会话列表（历史入口备用）。 */
export async function listStoryFlows(bookId: number): Promise<StoryFlow[]> {
  const { data } = await apiClient.get<{ items: StoryFlow[] }>('/story-flows/', {
    params: { bookId },
  });
  return data.items ?? [];
}

/** 删除剧情流会话（节点级联删除）。 */
export async function deleteStoryFlow(flowId: number): Promise<void> {
  await apiClient.delete(`/story-flows/${flowId}`);
}
